postgres/datadog_checks/postgres/statements.py

```diff
@@ -0,0 +1,191 @@
+    'temp_blks_read': 'postgresql.queries.temp_blks_read',
+    'temp_blks_written': 'postgresql.queries.temp_blks_written',
+}
+
+# Columns to apply as tags
+PG_STAT_STATEMENTS_TAG_COLUMNS = {
+    'datname': 'db',
+    'rolname': 'user',
+    'query': 'query',
+}
+
+DEFAULT_METRIC_LIMITS = {k: (100000, 100000) for k in PG_STAT_STATEMENTS_METRIC_COLUMNS.keys()}
+
+
+class PostgresStatementMetrics(object):
+    """Collects telemetry for SQL statements"""
+
+    def __init__(self, config):
+        self.config = config
+        # Cache results of monotonic pg_stat_statements to compare to previous collection
+        self._state = StatementMetrics()
+
+        # Available columns will be queried once and cached as the source of truth.
+        self._pg_stat_statements_columns = None
+        self._pg_stat_statements_query_columns = None
+
+    def _execute_query(self, cursor, query):
+        try:
+            cursor.execute(query)
+            return cursor.fetchall()
+        except (psycopg2.ProgrammingError, psycopg2.errors.QueryCanceled) as e:
+            logger.warning('Statement-level metrics are unavailable: %s', e)
+            return []
+
+    def init_pg_stat_statements_columns(self, db):
+        """
+        Load and cache the list of the columns available under the `pg_stat_statements` table. This must be
+        queried because version is not a reliable way to determine the available columns on `pg_stat_statements`.
+        The database can be upgraded without upgrading extensions, even when the extension is included by default.
+        """
+        if self._pg_stat_statements_columns is not None and self._pg_stat_statements_query_columns is not None:
+            return
+
+        query = """
+            SELECT column_name
+            FROM information_schema.columns
+            WHERE table_schema = 'public'
+            AND table_name = 'pg_stat_statements';
+            """
+        cursor = db.cursor()
+        columns = self._execute_query(cursor, query)
+        self._pg_stat_statements_columns = frozenset(column[0] for column in columns)
+
+        # Given all of the available columns, determine the columns that will actually be used to query
+        desired_columns = (
+            list(PG_STAT_STATEMENTS_METRIC_COLUMNS.keys())
+            + list(PG_STAT_STATEMENTS_OPTIONAL_COLUMNS)
+            + list(PG_STAT_STATEMENTS_TAG_COLUMNS.keys())
+        )
+        self._pg_stat_statements_query_columns = sorted(list(set(desired_columns) & self._pg_stat_statements_columns))
+
+    def collect_per_statement_metrics(self, instance, db, instance_tags):
+        if not self.config.deep_database_monitoring:
+            return
+
+        try:
+            self.__collect_per_statement_metrics(db, instance_tags)
+        except Exception:
+            logger.exception('Unable to collect statement metrics due to an error')
+
+    def __collect_per_statement_metrics(self, instance, db, instance_tags):
```

Comment:
This was to wrap in a try-except (see the method above) and indicate that this is not the method that should be called, but the wrapper. Because this is a new feature, there is a risk of an unexpected exception which would block the service check from running (something a lot of customers--like org2--page on).